bbbc021_feature_reduction_tests.py

```diff
@@ -0,0 +1,247 @@
+        if flag:
+            all_image_features = all_image_features.append(pd.DataFrame(feature_names).transpose())
+            flag = False
+        image_features = pd.DataFrame(image_features).transpose()
+        all_image_features = all_image_features.append(image_features, ignore_index=True)
+        print("Image processed: {} out of {}".format(i, len(filenames)))
+        i += 1
+    # concatenate filenames column to the features and save to CSV.
+    all_image_features = pd.concat([filenames_col, all_image_features], axis=1)
+    all_image_features.to_csv(out_file)
+
+
+def pca_transform(features_filename):
+    """
+    Read a file of image features into dataframe and perform a 2 component PCA, returning the 2 component values
+    of each image
+
+    :param features_filename: filename of CSV containing image features
+    :return coords: np array of 2 components for each image
+    """
+    all_image_features = pd.read_csv(features_filename)
+    pca = PCA(2)
+    coords = pca.fit_transform(all_image_features.iloc[1:, 2:])
+
+    return coords
+
+def tsne_transform(features_filename):
+    """
+    Read a file of image features into dataframe and perform TSNE transformation, returning the 2 component values
+    of each image
+
+    :param features_filename: filename of CSV containing image features
+    :return coords: np array of 2 components for each image
+    """
+    all_image_features = pd.read_csv(features_filename)
+
+    return TSNE().fit_transform(all_image_features.iloc[1:, 2:])
+
+
+def umap_transform(features_filename):
+    """
+    Read a file of image features into dataframe and perform umap transformation, returning the 2 component values
+    of each image
+
+    :param features_filename: filename of CSV containing image features
+    :return coords: np array of 2 components for each image
+    """
+    all_image_features = pd.read_csv(features_filename)
+
+    return umap.UMAP().fit_transform(all_image_features.iloc[1:, 2:])
+
+def generate_bokeh_csv(coords, names, filename):
+    """
+    Generate a CSV of columns
+        index,info,url,x,y
+    to work with Bokeh app.
+
+    :param coords: the x,y components of each data point
+    :param names: the names of the images you wish to load into bokeh, relative to the top level directory
+    """
+    coords_df = pd.DataFrame(coords)
+    filename_reg = r'(/.*/.*/)(Week.*_...)(.*_montaged\.tif)$'
+    indices = []
+    info = []
+    urls = []
+    
+    for name in names:
+        match = re.search(filename_reg, name)
+        if match:
+            indices.append(match.group(2))
+            info.append(match.group(2) + "_info")
+            urls.append(match.group(2) + match.group(3))
+
+    indices = pd.DataFrame(indices)
+    info = pd.DataFrame(info)
+    urls = pd.DataFrame(urls)
+
+    coord_csv = pd.concat([indices, info, urls, coords_df], axis=1)
+    coord_csv.columns = ["index", "info", "url", "x", "y"]
+    coord_csv.to_csv(filename)
+
+main()
```

Comment:
Suggest using:

```python
if __name__ == '__main__':
    main()
```

to allow re-use of functions by importing them into other scripts. Might be worth the extra flexibility.